### PR TITLE
Add DEV_GUIDE.md

### PR DESCRIPTION
**Pull request by TaekyungHeo**
## Summary
Add DEV_GUIDE.md

## Test Plan
Not needed.

### CODE REVIEW

File: DEV_GUIDE.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        "NcclTest",
+        "NcclTestGradingStrategy",
+        "NcclTestReportGenerationStrategy",
+        "NcclTestSlurmCommandGenStrategy",
+        "NcclTestSlurmInstallStrategy",
+        ...
+    ]
+    ```
+
+### Testing Procedures
```
Comment (amaslenn) [inline DEV_GUIDE.md:120]: Do you think we can make it a `pytest`?

File: DEV_GUIDE.md [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+      --agg_iters 1 \
+      --average 1 \
+      --parallel_init 0 \
+      --check 1 \
+      --blocking 0 \
+      --cudagraph 0 \
+      --stepfactor 2
+   ```
+
+## Setting Up the Test Template in CloudAI
```
Comment (amaslenn) [inline DEV_GUIDE.md:50]: IMO we should start with a higher level overview and focus on implementing custom flows similar to our CLI:

```markdown
# Implementing a custom execution flow

```py
# Create a System, TestTemplate and TestScenario objects
system, test_templates, test_scenario = System(), [TestTemplate()], TestScenario()

# One way to do that is to use the Parser class
parser = Parser(<system_config_path>, <test_template_path>, <test_path>, <test_scenario_path>)
system, test_templates, test_scenario = parser.parse()

# Update the system object using relevant SystemObjectUpdater. This is necessary to update the system object
# with the correct system configuration.
system_object_updater = SystemObjectUpdater()
system_object_updater.update(system)

# Check if test templates are installed
installer = Installer(system)
if not installer.is_installed(test_templates):
    # raise an exception or
    installer.install(test_templates)

# Create a Runner object and run the test scenario. Use asyncio.run() to run the async function.
runner = Runner(args.mode, system, test_scenario)
asyncio.run(runner.run())

# Generate a report...
generator = ReportGenerator(runner.runner.output_path)
generator.generate_report(test_scenario)

# ... and grade the test scenario
grader = Grader(runner.runner.output_path)
grader.grade(test_scenario)

```

We might also need a class diagrams, but let's postpone that till we figure our the interface of Core API.

Comment (TaekyungHeo) [reply DEV_GUIDE.md:50]: It makes perfect sense. I have written this DEV_GUIDE.md to inform others on how to add a new test template. I understsand this is not complete, and we need more detailed guidelines and documentation. Let's improve the document gradually.

cc: @srinivas212

File: DEV_GUIDE.md (hunk 1 above)
Comment (TaekyungHeo) [reply DEV_GUIDE.md:120]: Maybe I am not experienced enough to set it up as a pytest. I have several concerns regarding testing these with pytest. Installation involves interaction with the file system and remote docker repositories. Running involves interaction with the cluster. Therefore, with pytest, it must be challenging to validate. Others will work with pytest.

File: DEV_GUIDE.md (hunk 2 above)
Comment (TaekyungHeo) [reply DEV_GUIDE.md:50]: Sharing the discussion that I had with Andrei via a call. Let's postpone the developer documentation for a while. While it's important to have it, we need additional refactoring and design updates. Let's keep it high level.

**Pull request closed without merging**